FV3/coarse_graining/coarse_graining.F90

```diff
@@ -513,4 +523,311 @@ subroutine mask_area_weights(area, phalf, upsampled_coarse_phalf, masked_area_we
+    coarse, mask, nz)
+    integer, intent(in) :: nz
+    real, intent(in) :: fine(is:ie,js_coarse:je_coarse+1,1:nz)
+    real, intent(in) :: dx(is:ie,js:je+1)
+    logical, intent(in) :: mask(is:ie,js_coarse:je_coarse+1,1:nz)
+    real, intent(out) :: coarse(is_coarse:ie_coarse,js_coarse:je_coarse+1,1:nz)
+
+    real, allocatable :: weights(:,:), downsampled_dx(:,:)
+
+    integer :: i, j, k, a, i_coarse, j_coarse
+
+    allocate(weights(is:ie,js_coarse:je_coarse+1))
+    allocate(downsampled_dx(is:ie,js_coarse:je_coarse+1))
+
+    downsampled_dx = dx(:,js:je+1:coarsening_factor)
+
+    a = coarsening_factor - 1
+    do k = 1, nz
+        where (mask(:,:,k))
+          weights = downsampled_dx
+        elsewhere
+          weights = 0.0
+        endwhere
+        do i = is, ie, coarsening_factor
+          i_coarse = (i - 1) / coarsening_factor + 1
+          do j = js, je + 1, coarsening_factor
+              j_coarse = (j - 1) / coarsening_factor + 1
+              coarse(i_coarse,j_coarse,k) = sum(weights(i:i+a,j_coarse) * fine(i:i+a,j_coarse,k)) / sum(weights(i:i+a,j_coarse))
+          enddo
+        enddo
+    enddo
+   end subroutine weighted_block_edge_average_x_pre_downsampled_masked
+
+   subroutine upsample_d_grid_x(field_in, field_out, nz)
+    integer, intent(in) :: nz
+    real, intent(in) :: field_in(is_coarse:ie_coarse,js_coarse:je_coarse+1,1:nz)
+    real, intent(out) :: field_out(is:ie,js_coarse:je_coarse+1,1:nz)
+
+    integer :: i, j, k, a, i_coarse
+    a = coarsening_factor - 1
+    do i = is, ie, coarsening_factor
+       i_coarse = (i - 1) / coarsening_factor + 1
+       do j = js_coarse, je_coarse + 1
+          do k = 1, nz
+             field_out(i:i+a,j,k) = field_in(i_coarse,j,k)
+          enddo
+       enddo
+    enddo
+   end subroutine upsample_d_grid_x
+
+   subroutine remap_edges_along_x(field, phalf, dx, ptop, result)
+    real, intent(in) :: field(is:ie,js:je+1,1:npz)
+    real, intent(in) :: phalf(is-1,ie+1,js-1,je+1,1:npz+1)
+    real, intent(in) :: dx(is:ie,js:je+1)
+    real, intent(in) :: ptop
+    real, intent(out) :: result(is_coarse:ie_coarse,js_coarse:je_coarse+1,1:npz)
+
+    real, allocatable, dimension(:,:,:) :: phalf_d_grid, coarse_phalf_d_grid, coarse_phalf_d_grid_on_fine, remapped
+    logical, allocatable :: mask(:,:,:)
+
+    integer :: i, i_coarse, j, j_coarse, k, kn, km, kord, iv
+
+    allocate(phalf_d_grid(is:ie,js_coarse:je_coarse+1,1:npz+1))
+    allocate(coarse_phalf_d_grid(is_coarse:ie_coarse,js_coarse:je_coarse+1,1:npz+1))
+    allocate(coarse_phalf_d_grid_on_fine(is:ie,js_coarse:je_coarse+1,1:npz+1))
+    allocate(remapped(is:ie,js_coarse:je_coarse+1,1:npz))
+    allocate(mask(is:ie,js_coarse:je_coarse+1,1:npz))
+
+    ! Hard-code parameters related to mappm.
+    kn = npz
+    km = npz
+    kord = 1
+    iv = 1
+
+    ! 1. Interpolate and downsample phalf
+    call a_grid_to_d_grid_y(phalf, phalf_d_grid, npz+1)  ! TODO: better name for subroutine
```

Comment:
TODO still outstanding?

Reply:
Indeed this could still use a better name.  Perhaps something like `interpolate_to_d_grid_and_downsample_along_y` would be an improvement.